// @condor-labs/logger/src/logger/index.js
const DEBUG = 'DEBUG',
    INFO = 'INFO',
    WARNING = 'WARNING',
    ERROR = 'ERROR';
const self = {
    debug: (msg) => self._log(msg, DEBUG),
    log: (msg) => self._log(msg, INFO),
    information: (msg) => self._log(msg, INFO),
    info: (msg) => self._log(msg, INFO),
    warning: (msg) => self._log(msg, WARNING),
    warn: (msg) => self._log(msg, WARNING),
    error: (msg) => self._log(msg, ERROR),
    err: (msg) => self._log(msg, ERROR),
    _log: (msg, level) => {
        // check if this message is about an Error
        if (msg instanceof Error) {
            msg = {
                code: msg.code,
                message: msg.message,
                stack: msg.stack
            };
        }
        // check if this is a Json
        if (typeof msg === 'object') {
            msg = {
                level,
                ...msg
            }
            msg = JSON.stringify(msg);
        } else {
            msg = JSON.stringify({ level, message: msg });
        }
        // send to stdout
        switch (level) {
            case ERROR: console.error(msg); break;
            case WARNING: console.warn(msg); break;
            case INFO: console.info(msg); break;
            case DEBUG: console.log(msg); break;
            default: console.log(msg); break;
        }
    }
};

module.exports = self;
